### PR TITLE
Angel develops episode 4

### PR DESCRIPTION
**Pull request by n400peanuts**

### CODE REVIEW

File: episodes/04-transformers.md [hunk 1]
```diff
@@ -0,0 +1,737 @@
...
+```
+Output:
+
+`torch.Size([1, 21])`
+
+The output is a 2-dimensional tensor where the first dimention contains 1 element (this dimension represents the batch size), and the second dimension contains 21 elements which are equivalent to the 21 tokens that BERT generated with our string input.
+
+::: callout
+
+You noticed tin the previous outputs the `tensor()` and `torch()` wrappers around the arrays of integers. This is showing that the `transformers` library uses `pytorch` underneath, one of the most popular libraries for Deep Learning in Python. Pytorch's basic unit is the Tensor.
```
Comment (ThijsVroegh) [inline episodes/04-transformers.md:123]:
```suggestion
You noticed in the previous outputs the `tensor()` and `torch()` wrappers around the arrays of integers. This is showing that the `transformers` library uses `pytorch` underneath, one of the most popular libraries for Deep Learning in Python. Pytorch's basic unit is the Tensor.
```

File: episodes/04-transformers.md [hunk 2]
```diff
@@ -0,0 +1,737 @@
...
+
+The Transformer is a neural network architecture proposed by Google researchers [in 2017](https://arxiv.org/pdf/1706.03762) to address these and other limitations of RNNs and LSTMs. In their paper, *Attention is all you Need*, they tackled especifically the problem of Machine Translation (MT), which in NLP terms is stated as: how to generate a sentence (sequence of words) in target language B given a sentence in source language A? In order to translate, first one neural network needs to _encode_ the meaning of the source language A into vector representations, and then a second neural network needs to _decode_ that representation into tokens that are understandable in language B. Therefore translation is modeling language B _conditioned_ on what language A originally said.
```
Comment (ThijsVroegh) [inline episodes/04-transformers.md:36]:
```suggestion
The Transformer is a neural network architecture proposed by Google researchers [in 2017](https://arxiv.org/pdf/1706.03762) to address these and other limitations of RNNs and LSTMs. In their paper, *Attention is all you Need*, they tackled specifically the problem of Machine Translation (MT), which in NLP terms is stated as: how to generate a sentence (sequence of words) in target language B given a sentence in source language A? In order to translate, first one neural network needs to _encode_ the meaning of the source language A into vector representations, and then a second neural network needs to _decode_ that representation into tokens that are understandable in language B. Therefore translation is modeling language B _conditioned_ on what language A originally said.
```

File: episodes/04-transformers.md [hunk 3]
```diff
@@ -0,0 +1,737 @@
...
+- **Next Sentence Prediction:** the Encoder gets a linear binary classifier on top, which is trained to decide for each pair of sequences A and B, if sequence A precedes sequence B in a text. For the sentence pair: "Maria loves Groningen." and "This is a city in the Netherlands." the output of the classifier is "True" and for the pair "Maria loves Groningen." and "It was a tasty cake." the output should be "false" as there is no obvious continuation between the two sentences.
+
+Already the second pre-training task gives us an idea of the power of BERT: after it has been pretrained on hundreds of thoudands of texts, one can plug-in a classifier on top and re-use the *linguistic* knowledge previously acquired to fine-tune it for a specific task, without needing to learn the weights of the whole network from scratch all over again. In the next sections we will describe the components of BERT and show how to use it. This model and hundreds of related transformer-based pre-trained encoders can also be found on [Hugging Face](https://huggingface.co/google-bert/bert-base-cased).
```
Comment (ThijsVroegh) [inline episodes/04-transformers.md:56]:
```suggestion
Already the second pre-training task gives us an idea of the power of BERT: after it has been pretrained on hundreds of thousands of texts, one can plug-in a classifier on top and re-use the *linguistic* knowledge previously acquired to fine-tune it for a specific task, without needing to learn the weights of the whole network from scratch all over again. In the next sections we will describe the components of BERT and show how to use it. This model and hundreds of related transformer-based pre-trained encoders can also be found on [Hugging Face](https://huggingface.co/google-bert/bert-base-cased).
```

File: episodes/04-transformers.md [hunk 4]
```diff
@@ -0,0 +1,737 @@
...
+
+1. **Tokenizer:** splits text into tokens that the model recognizes
+2. **Embedder:** converts each token into a fixed-sized vector that represents it. These vectors are the actual input for the Encoder.
+3. **Encoder** several neural layers that model the token-level interactions of the input sequence to enhance meaning representation. The output of the encoder is a set of **H**idden layers, the vector representation of the ingested sequence.
+5. **Output Layer:** the final encoder layer (which we depict as a sequence **H**'s in the figure) contains arguably the best token-level representations that encode syntactic and semantic properties of each token, but this time each vector is already contextualized with the  the specific sequence.
```
Comment (ThijsVroegh) [inline episodes/04-transformers.md:68]:
```suggestion
5. **Output Layer:** the final encoder layer (which we depict as a sequence **H**'s in the figure) contains arguably the best token-level representations that encode syntactic and semantic properties of each token, but this time each vector is already contextualized with the specific sequence.
```

File: episodes/04-transformers.md [hunk 5]
```diff
@@ -0,0 +1,737 @@
...
+model = BertModel.from_pretrained("bert-base-cased")
+```
+
+## BERT Tokenizer
+
+We start with a string of text as written in any blog, book, newspaper etcetera. The `tokenizer` object is responsible of splitting the string into recognizable tokens for the model and embedding the tokens into their vector representations
+
+```python
+text = "Maria's passion for music is clearly heard in every note and every enchanting melody."
+# text = "Maria loves Groningen"
```
Comment (ThijsVroegh) [inline episodes/04-transformers.md:95]:
```suggestion
```

File: episodes/04-transformers.md [hunk 6]
```diff
@@ -0,0 +1,737 @@
...
+
+`torch.Size([1, 21])`
+
+The output is a 2-dimensional tensor where the first dimention contains 1 element (this dimension represents the batch size), and the second dimension contains 21 elements which are equivalent to the 21 tokens that BERT generated with our string input.
+
+::: callout
+
+You noticed tin the previous outputs the `tensor()` and `torch()` wrappers around the arrays of integers. This is showing that the `transformers` library uses `pytorch` underneath, one of the most popular libraries for Deep Learning in Python. Pytorch's basic unit is the Tensor.
+
+A *tensor* is a generalization for multidimentional array data. By convention, a vector is a 1-dimensional sequence of scalar numbers (or a 1-dim tensor), a matrix is a 2-dimensional sequence (2-dim tensor) and for N-dimensions where N > 2 we use the concept of tensor.
```
Comment (ThijsVroegh) [inline episodes/04-transformers.md:125]:
```suggestion
A *tensor* is a generalization of a multidimentional array of data. By convention, a vector is a 1-dimensional sequence of scalar numbers (or a 1-dim tensor), a matrix is a 2-dimensional sequence (2-dim tensor) and for N-dimensions where N > 2 we use the concept of tensor.
```

File: episodes/04-transformers.md [hunk 7]
```diff
@@ -0,0 +1,737 @@
...
+    print(output)
+    print(output.last_hidden_state.shape)
+```
+
+:::
+
+
+# BERT as a Language Model
+
+Now that we know hoe to embedd and run the model for obtaining the representations we can test the code for our first NLP Task: Language Modelling (LM). As mentioned before, the main pre-training task of BERT is, LM; in general this is the task of calculating the probability of a word based on the known neighboring words (yes, Word2Vec was also a kind of LM). Obtaining training data for this task is very cheap, as all we need is millions of sentences from existing texts, without any labels. In this setting, BERT encodes a sequence of words, and predicts from a set of English tokens, what is the most likely token that could be inserted in the `[MASK]` position
```
Comment (ThijsVroegh) [inline episodes/04-transformers.md:198]:
```suggestion
Now that we know how to embedd and run the model to obtain the representations, we can test the code for our first NLP Task: Language Modelling (LM). As mentioned before, the main pre-training task of BERT is LM: calculating the probability of a word based on the known neighboring words (yes, Word2Vec was also a kind of LM). Obtaining training data for this task is very cheap, as all we need is millions of sentences from existing texts, without any labels. In this setting, BERT encodes a sequence of words, and predicts from a set of English tokens, what is the most likely token that could be inserted in the `[MASK]` position
```

File: episodes/04-transformers.md [hunk 8]
```diff
@@ -0,0 +1,737 @@
...
+=====	 Maria [MASK] Groningen
+{'score': 0.24399833381175995, 'token': 117, 'token_str': ',', 'sequence': 'Maria, Groningen'}
+{'score': 0.12300779670476913, 'token': 1104, 'token_str': 'of', 'sequence': 'Maria of Groningen'}
+{'score': 0.11991506069898605, 'token': 1107, 'token_str': 'in', 'sequence': 'Maria in Groningen'}
+{'score': 0.07722211629152298, 'token': 1306, 'token_str': '##m', 'sequence': 'Mariam Groningen'}
+{'score': 0.0632941722869873, 'token': 118, 'token_str': '-', 'sequence': 'Maria - Groningen'}
+
+```
+
+When we call the `nlp` pipeline, requesting to return the `top_k` most likely suggestions to complete the provided sentences (in this case `k=5`). The pipeline returns a list of outputs as python dictionaries. Depending on the task, the fields of the dictionary will differ. In this case, the `fill-mask` task returns a score (between 0 and 1, the higher the score the more likely the token is), a tokenId, and its corresponding string, as well as the full "unmasked" sequence.
```
Comment (ThijsVroegh) [inline episodes/04-transformers.md:247]:
```suggestion
When we call the `nlp` pipeline, we request to return the `top_k` most likely suggestions which complete the provided sentences (in this case `k=5`). The pipeline returns a list of outputs as python dictionaries. Depending on the task, the fields of the dictionary will differ. In this case, the `fill-mask` task returns a *probability score* (between 0 and 1: the higher the score, the more likely the token is), a *tokenId* and its corresponding string, as well as the full "unmasked" sequence. 
```

File: episodes/04-transformers.md [hunk 9]
```diff
@@ -0,0 +1,737 @@
...
+![BERT as an Emotion Classifier](fig/bert4.png)
+
+
+Let's see the example of a ready pre-trained emotion classifier based on `RoBERTa` model. This model was fine-tuned in the Go emotions [dataset](https://huggingface.co/datasets/google-research-datasets/go_emotions), taken from English Reddit and labeled for 28 different emotions at the sentence level. The fine-tuned model is called [roberta-base-go_emotions](https://huggingface.co/SamLowe/roberta-base-go_emotions). This model takes a sentence as input and ouputs a probability distribution over the 28 possible emotions that might be conveyed in the text. For example:
+
+```python
+
+classifier = pipeline(task="text-classification", model="SamLowe/roberta-base-go_emotions", top_k=3)
+
+sentences = ["I am not having a great day", "Maria loves Groningen"]
```
Comment (ThijsVroegh) [inline episodes/04-transformers.md:268]: Add "This is a lovely and innocent sentence"

File: episodes/04-transformers.md [hunk 10]
```diff
@@ -0,0 +1,737 @@
...
+
+The next step is to give the sequence of tokens to the Encoder which processes it through the transformer layers and outputs a sequence of dense vectors:
+
+```python
+with torch.no_grad():
+    output = model(**encoded_input)
+    print(output.last_hidden_state[0])
+```
+
+When we print the vectors we only see a lot of fine-tuned weights which are not very informative in their own, but the full-vectors are meaningful within the embedding space, which emulates some aspects of linguistic meaning. In the case of wanting to obtain a single vector for *enchanting*, you can average the three vectors that belong to the token pieces that ultimately from that word.
```
Comment (ThijsVroegh) [inline episodes/04-transformers.md:336]:
```suggestion
When we print the vectors we only see a lot of fine-tuned weights which are not very informative in their own, but the full-vectors are meaningful within the embedding space, which emulates some aspects of linguistic meaning. In the case of wanting to obtain a single vector for *enchanting*, you can average the three vectors that belong to the token pieces that ultimately form that word.
```

File: episodes/04-transformers.md [hunk 11]
```diff
@@ -0,0 +1,737 @@
...
+```python
+# Search for the index of 'note' and obtain its vector from the sequence
+note_index_1 = string_tokens.index('note')
+note_vector_1 = output.last_hidden_state[0][note_index_1].detach().numpy()
+note_token_id_1 = token_ids[note_index_1]
+
+print(note_index_1, note_token_id_1, string_tokens)
+print(note_vector_1[:5])
+```
+We are basically printing the tokenized sentence from the previous exaple and shoging the index of the token `note` in the list of tokens. We are also printing the tokenID assigned to this token and the list of tokens. Finally, the last print shows the first five dimensions of the vector representing the token `note`.
```
Comment (ThijsVroegh) [inline episodes/04-transformers.md:350]:
```suggestion
We are basically printing the tokenized sentence from the previous example and showing the index of the token `note` in the list of tokens. We are also printing the tokenID assigned to this token and the list of tokens. Finally, the last print shows the first five dimensions of the vector representing the token `note`.
```

File: episodes/04-transformers.md [hunk 12]
```diff
@@ -0,0 +1,737 @@
...
+
+
+## The Attention Mechanism
+
+The original attention mechanism (remember this was developed for language translation) is a component in between the Encoder and the Decoder that helps the model to _align_ the important information from the input sequence in order to generate a more accurate token in the output sequence:
+
+
+![The Encoder-Decoder Attention Mechanism](fig/trans3.png)
+
+In the example above, the attention puts more weight in the input _etudiant_, so the decoder uses that information to _know_ that is should generate _student_. Note that if the decoder based it's next work probability just based in the sequence "I am a ..." it could basically generate any word and still sound natural, but it is thanks to the attention mechanism that it preserves the meaning of the input sequence.
```
Comment (ThijsVroegh) [inline episodes/04-transformers.md:401]: there is a mismatch between the example sentence in the image and the text (which uses French)

Comment (ThijsVroegh) [inline episodes/04-transformers.md:401]:
```suggestion
In the example above, the attention puts more weight in the input _etudiant_, so the decoder uses that information to _know_ that is should generate _student_. Note that if the decoder based it's next word probability just on the sequence "I am a ...", it could basically generate any word and still sound natural. However, it is thanks to the attention mechanism that it preserves the meaning of the input sequence.
```

**Review by ThijsVroegh — COMMENTED** [review]
Hi @n400peanuts and @angel-daza, I went through the entire episode, which i really liked reading! Thanks, impressive work!

There are many explanatory small explanations/ experiments clarifying the underlying mechanisms of transformers, really helping to better understand what is happening (also helpful for understanding LLMs). 

In addition: 
•	The explanations and descriptions are technically accurate and cover essential concepts. 
•	Overall, I think the material is well-balanced up to the END HERE mark
•	The code examples are very appropriate and simple enough for illustrating the sometimes complex concepts, and demonstrate practical usage.
•	There are some minor typos and formatting issues that still need to be addressed.
•	I’m not sure if all the material can be covered within the designated time, though, but a practice session will make this clear I suppose. The END HERE mark is a good idea to already have in mind beforehand what to skip in case there is indeed too little time. 
•	Finally, a couple of minor remarks:
-	Consistent use of quotes: Ensure consistent use of single or double quotes for titles and other text elements.
-	Image descriptions: Provide more meaningful descriptions below the images. Instead of e.g., "Transformer Architecture," describe what the image depicts in more detail in ca. 2 sentences, even if it repeats the explanation in the text .
-	Code snippets: Ensure code snippets are properly formatted and include comments to explain each step.
-	Readability and flow: Break down long paragraphs into shorter ones. Use bullet points or numbered lists for steps or key points to make the text easier to scan.
-	Clarify any technical term: Explain any technical terms or abbreviations when they are first introduced (e.g. , LSTM).
-	Use of callouts: I really like these callouts for further background information. Please utilize these callouts a bit more to highlight important information, even if it appears redundant. For example, what is softmax?
-	Grammar and spelling: Proofread for grammatical errors and typos. For example, "tin" should be "in," and "shoging" should be "showing." I already gave some suggestions which you can directly commit in case you agree with them.
-	The comment on Finetunning BERT being very cheap is interesting! Could you perhaps, if possible and not too complicated, add a code snippit in the call out for those interested in applying this finetuning?
-	I still find the concept of self-attention a bit difficult as it is explained now, although the example wit the ‘it’ referencing to a particular noun is very clear.

File: episodes/04-transformers.md (hunk 8 above)
Comment (angel-daza) [reply episodes/04-transformers.md:247]: I prefer to avoid using "probability" because, strictly speaking, this is not a probability distribution nor a confidence score, it just happens to be between 0 and 1 and is conveniently interpreted as Prob...
